src/openqaoa-core/algorithms/baseworkflow.py

```diff
@@ -0,0 +1,578 @@
+#   limitations under the License.
+
+from abc import ABC
+from typing import List
+import json
+import gzip
+from os.path import exists
+
+from .workflow_properties import BackendProperties, ClassicalOptimizer
+from ..backends import DeviceLocal, DeviceBase
+from ..problems import QUBO
+from ..utilities import delete_keys_from_dict, is_valid_uuid, generate_uuid
+from ..backends.qaoa_backend import (
+    DEVICE_NAME_TO_OBJECT_MAPPER,
+    DEVICE_ACCESS_OBJECT_MAPPER,
+)
+
+
+class Workflow(ABC):
+    """
+    Abstract class to represent an optimizer
+
+    It's basic usage consists of
+
+     #. Initialization
+     #. Compilation
+     #. Optimization
+
+    Attributes
+    ----------
+    device: `DeviceBase`
+        Device to be used by the optimizer
+    backend_properties: `BackendProperties`
+        The backend properties of the optimizer workflow. Use to set the
+        backend properties such as the number of shots and the cvar values.
+        For a complete list of its parameters and usage please see the method `set_backend_properties`
+    classical_optimizer: `ClassicalOptimizer`
+        The classical optimiser properties of the optimizer workflow.
+        Use to set the classical optimiser needed for the classical optimisation part of the optimizer routine.
+        For a complete list of its parameters and usage please see the method `set_classical_optimizer`
+    local_simulators: `list[str]`
+        A list containing the available local simulators
+    cloud_provider: `list[str]`
+        A list containing the available cloud providers
+    compiled: `Bool`
+        A boolean flag to check whether the optimizer object has been correctly compiled at least once
+    id: TODO
+    exp_tags: `dict`
+        A dictionary containing the tags of the optimizer object. The user can set this
+        value using the set_exp_tags method.
+    problem: `Problem`
+        The problem object that the optimizer will be optimizing.
+    results: `Results`
+        The results object that will contain the results of the optimization routine.
+    """
+
+    def __init__(self, device=DeviceLocal("vectorized")):
+        """
+        Initialize the optimizer class.
+
+        Parameters
+        ----------
+        device: `DeviceBase`
+            Device to be used by the optimizer. Default is using the local 'vectorized' simulator.
+        """
+
+        self.device = device
+        self.backend_properties = BackendProperties()
+        self.classical_optimizer = ClassicalOptimizer()
+        self.local_simulators = list(DEVICE_NAME_TO_OBJECT_MAPPER.keys())
+        self.cloud_provider = list(DEVICE_ACCESS_OBJECT_MAPPER.keys())
+        self.compiled = False
+
+        # Initialize the identifier stamps, we initialize all the stamps needed to None
+        self.header = {
+            "atomic_uuid": None,  # the uuid of the run it is generated automatically in the compilation
```

Comment:
Can we change `uuid` to `id`? We should have done it earlier, but we lost it somewhere cc @raulconchello

Reply:
then should we change also the functions `is_valid_uuid`, `generate_uuid` in utilities should also change to `is_valid_id`, `generate_id`

Reply:
Sorry, I missed this comment. No, no need to change that. It is actually good for coders to remember that those are `uuid4`